### PR TITLE
Add gRPC filter

### LINKED ISSUE #373: Filter out health check request in grpc
**Is your feature request related to a problem?**
As of 0.18b1, `opentelemetry-instrumentation-grpc` doesn't have the feature to exclude trace propagation to the specific endpoints. Because of this, gRPC servers always send health check request traces to the backend via the exporter as you see in the screenshot below:
image

**Describe the solution you'd like**
It would be nice if opentelemetry-instrumentation-grpc could have something similar to `OTEL_PYTHON_FLASK_EXCLUDED_URLS` in `opentelemetry-instrumentation-flask`.

**Describe alternatives you've considered**
Each exporters support filter option to avoid sending unnecessary traces to the backend.

**Additional context**
Health checking is the common practice in Kubernetes and the traces for such requests sometimes pollute the Trace dashboard and leads to unnecessary costs for SaaS services.

### PR DESCRIPTION
**Pull request by ymotongpoo**
# Description

This pull request adds section to handle filter in client and server interceptors and provides pre-defined commonly used filters.

Fixes #373

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] tox -e py39-test-instrumentation-grpc (test_filters.py and test_xxx_filters.py)

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by aabmass** [comment]
Thanks for the PR! In the issue you mentioned environment variables, something like `OTEL_PYTHON_FLASK_EXCLUDED_URLS` but for gRPC. These are nice for auto instrumentation. Is the plan to implement similar environment variables on top of these predicates? Is there an easy way to specify the filters for auto instrumentation?

**Comment by ymotongpoo** [comment]
@aabmass thanks for the comment! though I tried to [look up the implementation](https://github.com/search?q=org%3Aopen-telemetry+OTEL_PYTHON_FLASK_EXCLUDED_URLS&type=code) for `OTLE_PYTHON_FLASK_EXCLUDED_URLS`, I couldn't find the point where it is implemented. Could you tell the pointer where the environment variable is handled?

**Comment by ymotongpoo** [comment]
Ok, now I understand [this part](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/cbf005be6fb17f35b4cbaa4e76ac30bb64b3a258/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L170) is referencing the environment variable via [`get_excluded_urls`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/cbf005be6fb17f35b4cbaa4e76ac30bb64b3a258/util/opentelemetry-util-http/src/opentelemetry/util/http/__init__.py#L80-L88). I think this function introduces unnecessary complexity into the repo, tho.

**Comment by ymotongpoo** [comment]
@aabmass ok, I added the support in the global interceptor to exclude requests to services specified in `OTEL_PYTHON_GRPC_EXCLUDED_SERVICES` in comma separated format.

**Comment by ymotongpoo** [comment]
@ocelotl @lzchen @srikanthccv it seems "auto assign owners" is broken and this PR is not assigned to reviewers properly. could anyone take a look?

**Comment by srikanthccv** [comment]
There are no owners for this instrumentation so the contribution approvers are the reviewers.

**Comment by ymotongpoo** [comment]
@srikanthccv thanks for the answer. ok, then I wait for the approvers' review.

**Comment by ymotongpoo** [comment]
@srikanthccv Is there any way to rerun the tests? Though I ran the test locally with `tox -e`, the failures are not reproduced. I would like to confirm if this is just a flaky or there are fundamental difference between local testing and GitHub Actions besides OS environments.

Also are there any way to get the approver's review process started?

**Comment by aabmass** [comment]
Tox is the right way. They are probably flakes or maybe an already known issue. I'll review this PR :)

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:77]: can use `any()` here like my comment above

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:82]: I think `not()` or `negate()` would be better names here

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:62]:
```suggestion
    """Returns a filter function that returns True if any of filter functions
```

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:69]:
```suggestion
        A filter function that returns True if any of filter functions
```

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:87]: I think this was accidentally copied from below?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:91]: same comment

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:185]: I think "method" is copied from above, but should be "servicE"

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:209]: Should `filters` be singular -> `filter`? Also, please add type annotation for the param

**Review by aabmass — COMMENTED** [review]
Thanks for the PR, sorry for all the comments

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:79]: Thanks for the suggestion! Because `not` is reserved keyword, I take `negate`.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:209]: Though I wanted to use singular, `filter` is the name of builtin function and that's the reason why I use plural here. I would like to choose better naming if possible.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:257]: I was confused and incorrectly merged your suggestion above. Actually this is the feature that only intercept the requests that applies to the filter, not eliminate those. This behavior is the same as the one in opentelemetry-go-contrib.

- https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/google.golang.org/grpc/otelgrpc/grpctrace.go#L36-L39
- https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http#WithFilter

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:40]: Thank you! I added the type annotations and it worked.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:74]: Done!

**Review by ymotongpoo — COMMENTED** [review]
adopted the suggestions and replied to the questions

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py
Comment (aabmass) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: Ah good point, I think one option would be `filter_`. I'll leave it up to you tho

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: yeah, I used `filter_` first but I refrained from using underscore as I thought it's not recommended in this repo. If it's okay, I'm happy to replace the variable name as you say.

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: yeah, I used `filter_` first but I refrained from using underscore as I thought it's not recommended in this repo. If it's okay, I'm happy to replace the variable name as you say.

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: yeah, I used `filter_` first but I refrained from using underscore as I thought it's not recommended in this repo. If it's okay, I'm happy to replace the variable name as you say.

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: @aabmass I talked with my friend about this and I heard from him that SQLAlchemy is using underscore to avoid the conflict with both keywords and builtin.
https://github.com/sqlalchemy/sqlalchemy/blob/6a496a5f40efe6d58b09eeca9320829789ceaa54/lib/sqlalchemy/ext/serializer.py#L81

I would like to leave the decision to the approvers and maintainers, but I hope to have make this discussion to be the convention in this project, and the rules for it in the linter. This allows upcoming contributors can apply the guideline without confusion.

@srikanthccv WDYT?

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: FYI: flake8 seems to have a plugin to achieve it. 
https://pypi.org/project/flake8-builtins/

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: Can you prefix it with something like `service_filter` or similar? I am fine with underscore too.

Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/__init__.py:208]: @srikanthccv thank you for the comment. Because this may include more filters that should cover other than services, I'm leaning towards `filter_` at this momoent.

**Comment by ymotongpoo** [comment]
I updated the option name from `filters` to `filter_`. The tests kicked by GitHub Action are flaky and failing with arbitrary unexpected errors.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:99]: I may have made a bad comment before. But `name` here should be `str` I think?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:120]: same here, should be `str`?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:141]: same here, should be `str`?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:162]: same here, should be `str`?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py:183]: same here, should be `str`?

File: instrumentation/opentelemetry-instrumentation-grpc/tests/test_filters.py
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-grpc/tests/test_filters.py:48]: nit I would prefer [`@dataclass`](https://docs.python.org/3/library/dataclasses.html) over named tuple

**Review by aabmass — APPROVED** [review]
LGTM!

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py]: I also should have checked the type more carefully. Thank you for catching this.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py]: ditto

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py]: ditto

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py]: ditto

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/filters/__init__.py]: ditto

File: instrumentation/opentelemetry-instrumentation-grpc/tests/test_filters.py
Comment (ymotongpoo) [reply instrumentation/opentelemetry-instrumentation-grpc/tests/test_filters.py:48]: Either `collections.namedtuple` or `@dataclass` is fine. I aligned the implementation here with the one in `grpcext`.

https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py#L28-L40

**Pull request merged**
